api/metrics.go

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) 2024 The VeChainThor developers
+
+// Distributed under the GNU Lesser General Public License v3.0 software license, see the accompanying
+// file LICENSE or <https://www.gnu.org/licenses/lgpl-3.0.html>
+
+package api
+
+import (
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/vechain/thor/v2/metrics"
+)
+
+var (
+	metricHttpReqCounter  = metrics.CounterVec("api_request_count", []string{"path", "code", "method"})
+	metricHttpReqDuration = metrics.HistogramVec("api_duration_ms", []string{"path", "code", "method"}, metrics.BucketHTTPReqs)
+)
+
+// metricsResponseWriter is a wrapper around http.ResponseWriter that captures the status code.
+type metricsResponseWriter struct {
+	http.ResponseWriter
+	statusCode int
+}
+
+func newMetricsResponseWriter(w http.ResponseWriter) *metricsResponseWriter {
+	return &metricsResponseWriter{w, http.StatusOK}
+}
+
+func (m *metricsResponseWriter) WriteHeader(code int) {
+	m.statusCode = code
+	m.ResponseWriter.WriteHeader(code)
+}
+
+// metricsHandler is a middleware that records metrics for each request.
+func metricsHandler(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		now := time.Now()
+
+		mrw := newMetricsResponseWriter(w)
+		h.ServeHTTP(mrw, r)
+
+		url := strings.ReplaceAll(strings.TrimLeft(r.URL.Path, "/"), "/", "_") // ensure no unexpected slashes
+		metricHttpReqCounter.AddWithLabel(1, map[string]string{"path": url, "code": strconv.Itoa(mrw.statusCode), "method": r.Method})
```

Comment:
I like this direction, but I'm starting to think that perhaps it's too generic.

It captures `/` and `doc_stoplight-ui_` and any api request, which means tracking even bad urls.

It also captures `accounts_0x0000000000000000000000000000456E65726779` or `accounts_0x0000000000000000000000000000456E65726779_code`, which makes this have a super high cardinality for prometheus. (one instrument per request almost)

Ie. capture `accounts_details` and `accounts_inspect_clauses` instead of `accounts_0x0000000000000000000000000000456E65726779`, because then we have 2 instruments, one per endpoints vs potential infinite instruments.

Reply:
I made this setup in draupnir to match regex if it's any good:

```go
var (
	GetAccountsAddress    = "GET /accounts/{address}"
	PostAccounts          = "POST /accounts"
	PostAccountsAddress   = "POST /accounts/{address}"
	PostAccountsStar      = "POST /accounts/*"
	GetAccountsStorage    = "GET /accounts/{address}/storage"
	GetAccountsCode       = "GET /accounts/{address}/code"
	GetTransaction        = "GET /transactions/{id}"
	GetTransactionReceipt = "GET /transactions/{id}/receipt"
	PostTransactions      = "POST /transactions"
	GetBlocks             = "GET /blocks/{revision}"
	PostLogsEvent         = "POST /logs/event"
	PostLogsTransfer      = "POST /logs/transfer"
	GetNodeNetworkPeers   = "GET /node/network/peers"
	PostDebugTracers      = "POST /debug/tracers"
	PostDebugTracersCall  = "POST /debug/tracers/call"
	PostDebugStorageRange = "POST /debug/storage-range"
)

var urlConfigs = []urlConfig{
	// Account endpoints
	{GetAccountsAddress, regexp.MustCompile("^/accounts/(0x)?[A-Fa-f0-9]{40}(\\?.*)?$"), "GET"},
	{PostAccounts, regexp.MustCompile("^/accounts/*(\\?.*)?$"), "POST"},
	{PostAccountsAddress, regexp.MustCompile("^/accounts/(0x)?[A-Fa-f0-9]{40}(\\?.*)?$"), "POST"},
	{PostAccountsStar, regexp.MustCompile("^/accounts/\\*"), "POST"},
	{GetAccountsStorage, regexp.MustCompile("^/accounts/(0x)?[A-Fa-f0-9]{40}/storage(\\?.*)?$"), "GET"},
	{GetAccountsCode, regexp.MustCompile("^/accounts/(0x)?[A-Fa-f0-9]{40}/code(\\?.*)?$"), "GET"},
	// Transaction endpoints
	{GetTransaction, regexp.MustCompile("^/transactions/[^/]+(\\?.*)?$"), "GET"},
	{GetTransactionReceipt, regexp.MustCompile("^/transactions/[^/]+/receipt(\\?.*)?$"), "GET"},
	{PostTransactions, regexp.MustCompile("^/transactions(\\?.*)?$"), "POST"},
	// Block endpoints
	{GetBlocks, regexp.MustCompile("^/blocks/[^/]+(\\?.*)?$"), "GET"},
	// Logs endpoints
	{PostLogsEvent, regexp.MustCompile("^/logs/event(\\?.*)?$"), "POST"},
	{PostLogsTransfer, regexp.MustCompile("^/logs/transfer(\\?.*)?$"), "POST"},
	// Node endpoints
	{GetNodeNetworkPeers, regexp.MustCompile("^/node/network/peers(\\?.*)?$"), "GET"},
	// Debug endpoints
	{PostDebugTracers, regexp.MustCompile("^/debug/tracers(\\?.*)?$"), "POST"},
	{PostDebugTracersCall, regexp.MustCompile("^/debug/trace/call(\\?.*)?$"), "POST"},
	{PostDebugStorageRange, regexp.MustCompile("^/debug/storage-range(\\?.*)?$"), "POST"},
}
```

Reply:
@otherview What you wanted is the filtering and mapping feature, I think it can be done in a subsequent PR, wdyt? @darrenvechain 's code might be a good approach

Reply:
Yeah, we defo want a mapping feature because we want to have a fixed set of URLs we're taking metrics on.

I'm starting to think that it might be needed to update the PR. What happens if kostas enables metrics in the public nodes, particularly with all the bad urls and different requests the public nodes receive?

Reply:
Only thing is that my regex expressions only catch valid addresses/ txs. Not sure if want to capture bad requests too